### PR TITLE
LibVideo: Use threaded tile column decoding on Lagom

### PR DESCRIPTION
**Pull request by Zaggy1024**
To make this change, syntax elements must be counted in each thread and then summed up after all work is done to prevent race conditions in the counting.

This improves decoding for two-column videos by about 40%.

### CODE REVIEW

File: AK/Debug.h.in [hunk 1]
```diff
@@ -498,6 +498,10 @@
 #    cmakedefine01 UTF8_DEBUG
 #endif
 
+#ifndef VIDEO_WORKER_THREAD_DEBUG
+#cmakedefine01 VIDEO_WORKER_THREAD_DEBUG
```
Comment (caoimhebyrne) [inline AK/Debug.h.in:502]: Indent here:
```
#ifndef VIDEO_WORKER_THREAD_DEBUG
#    cmakedefine01 VIDEO_WORKER_THREAD_DEBUG
#endif
```

**Review by caoimhebyrne — CHANGES_REQUESTED** [review]
Didn't take a look at the actual code yet (may look at that later on), but just noticed one thing really quickly

**Comment by ADKaster** [comment]
I didn't look *too* closely but two things came to mind:

- WorkerThread looks like it belongs in LibThreading
- creating a new set of worker threads each time we parse a new video seems wasteful if we end up parsing multiple videos in one execution. Should the allocation of threads be deferred to some "ThreadPool" class that lets you submit to N threads? The pool would park the threads when they're not in use, waiting for a semaphore to be posted that lets them know there's a new item in the queue.

**Comment by Zaggy1024** [comment]
> * WorkerThread looks like it belongs in LibThreading

Seems reasonable, I'll look into moving it there.

> * creating a new set of worker threads each time we parse a new video seems wasteful if we end up parsing multiple videos in one execution. Should the allocation of threads be deferred to some "ThreadPool" class that lets you submit to N threads? The pool would park the threads when they're not in use, waiting for a semaphore to be posted that lets them know there's a new item in the queue.

I think that making a shared pool of decoder threads is more complexity than the decoder needs. The purpose of these workers is to do a single task at a time and allow a simple function call to wait for that task to complete, which works well for the decoder as it stands now. Throwing single tile column at a shared thread pool seems like it could result in the decode having to wait for another concurrently-running decoder to finish its tile column, which could potentially cause it to be delayed way longer than desirable.

However, at a later point I may need to look into setting a thread limit for each decoder, since it may not be good to as many threads at the problem as possible once the decoder is fast enough to run in real-time with less. It's not currently able to do that, though, so I want to defer that work.

**Comment by Zaggy1024** [comment]
I've moved WorkerThread to LibThreading, this should be ready for another look.

File: Userland/Libraries/LibThreading/WorkerThread.h [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    ~WorkerThread()
+    {
+        m_mutex.lock();
+        m_stop = true;
+        m_condition.broadcast();
+        while (!is_in_state(State::Stopped)) {
+            m_condition.wait();
+        }
```
Comment (GMTA) [inline Userland/Libraries/LibThreading/WorkerThread.h:89]: Same here

File: Userland/Libraries/LibVideo/VP9/Parser.cpp [hunk 3]
```diff
@@ -18,6 +19,12 @@
 #    pragma GCC optimize("O3")
 #endif
 
+// Threading is unstable in Serenity with smp=on, but is stable in Lagom.
+// Let's keep this disabled on Serenity for now.
+#if !defined(AK_OS_SERENITY)
+#    define VP9_TILE_THREADING
+#endif
```
Comment (GMTA) [inline Userland/Libraries/LibVideo/VP9/Parser.cpp:26]: I'd argue that exposing these bugs with SMP enabled will motivate more people to fix them :^) and SMP is disabled by default anyway.

File: Userland/Libraries/LibVideo/VP9/Parser.cpp [hunk 4]
```diff
@@ -885,12 +898,61 @@ DecoderErrorOr<void> Parser::decode_tiles(FrameContext& frame_context)
...
         }
     }
+
+    auto decode_tile_column = [this, tile_rows](auto& column_workloads) -> DecoderErrorOr<void> {
+        VERIFY(column_workloads.size() == tile_rows);
+        for (auto tile_row = 0u; tile_row < tile_rows; tile_row++) {
+            auto result = decode_tile(column_workloads[tile_row]);
+            if (result.is_error()) {
+                return result;
+            }
```
Comment (GMTA) [inline Userland/Libraries/LibVideo/VP9/Parser.cpp:912]: Can't this be a `TRY`?

File: Userland/Libraries/LibVideo/VP9/Parser.cpp (hunk 3 above)
Comment (Zaggy1024) [reply Userland/Libraries/LibVideo/VP9/Parser.cpp:26]: Fair enough, I think that the threading does incur a slight penalty with SMP off but it's probably fine. I'll be doing most of my testing in Lagom anyway :^)

File: Userland/Libraries/LibVideo/VP9/Parser.cpp (hunk 4 above)
Comment (Zaggy1024) [reply Userland/Libraries/LibVideo/VP9/Parser.cpp:912]: Good catch, thanks!

File: Userland/Libraries/LibVideo/VP9/Parser.cpp [hunk 5]
```diff
@@ -82,6 +81,9 @@ Vector<size_t> Parser::parse_superframe_sizes(ReadonlyBytes frame_data)
 /* (6.1) */
 DecoderErrorOr<FrameContext> Parser::parse_frame(ReadonlyBytes frame_data)
 {
+    if (!m_probability_tables) {
+        m_probability_tables = DECODER_TRY_ALLOC(try_make<ProbabilityTables>());
+    }
```
[the remark this answers is not in the document]
Comment (Zaggy1024) [reply Userland/Libraries/LibVideo/VP9/Parser.cpp]: Hopefully I got all of them in this PR. For some reason I had been under the impression blocks were preferred always, so I'll have to correct a few of those in my other PRs when I get the chance to go through them.

File: Userland/Libraries/LibVideo/VP9/Parser.cpp [hunk 6]
```diff
@@ -865,8 +869,15 @@ DecoderErrorOr<void> Parser::decode_tiles(FrameContext& frame_context)
     //        then run through each column of tiles in top to bottom order afterward. Each column can be sent to a worker thread
     //        for execution. Each worker thread will want to create a set of above contexts sized to its tile width, then provide
     //        those to each tile as it decodes them.
-    for (auto tile_row = 0; tile_row < tile_rows; tile_row++) {
-        for (auto tile_col = 0; tile_col < tile_cols; tile_col++) {
+    Vector<Vector<TileContext, 1>, 4> tile_workloads;
+    DECODER_TRY_ALLOC(tile_workloads.try_ensure_capacity(tile_cols));
+    for (auto tile_col = 0u; tile_col < tile_cols; tile_col++) {
+        tile_workloads.append({});
+        tile_workloads[tile_col].ensure_capacity(tile_rows);
```
Comment (GMTA) [inline Userland/Libraries/LibVideo/VP9/Parser.cpp:876]: Ayy you sneaked this in :^) shouldn't this be `try_ensure_capacity`?

Comment (Zaggy1024) [reply Userland/Libraries/LibVideo/VP9/Parser.cpp:876]: Oops, yeah, not awake enough today I suppose.

**Pull request merged**